packages/store/src/trieStore.js

```diff
@@ -0,0 +1,184 @@
+// is to compare to arrays element by element using normal
+// JavaScript map equality.
+// ```js
+// trieKeyEqual([NaN, 0], [NaN, -0]);
+// ```
+export const trieKeyEqual = (a, b) =>
+  a.length === b.length && a.every((v, i) => mapKeyEqual(v, b[i]));
+
+// The Pumpkin must not escape. It must be distinct from any value that
+// could be passed into this module.
+const Pumpkin = harden({});
+
+// Not intended for export; just for internal use.
+// `enumerable` should only be true if the `makeStore` argument
+// makes enumerable stores.
+function makeTrieStoreMaker(makeStore, enumerable = false) {
+  /**
+   * A trieStore is a *store* (as defined in store.js) whose keys are arrays
+   * that are compared and looked up based on element-by-element equality.
+   * Internally it uses a trie made of a tree of stores. We delegate the
+   * store error detection and handling to the underlying stores, as made by
+   * the `makeStore` argument. These underlying stores must not leak the
+   * Pumpkin.
+   * TODO This delegation is not right because the underlying stores will
+   * report the Pumpkin as the key that was or was not present.
+   *
+   * @param  {string} keyName - the column name for the key
+   */
+  function makeTrieStore(keyName = 'key') {
+    const root = makeStore(keyName);
+    // This is skeaky. This caches lastStore keyed by the key *identity*.
+    // That way, once a given key is dropped, the cache entry gets collected.
+    // If the same key identity is reused, such as by a .has followed by a
+    // .get, it's fast. Should have no observable effect.
+    //
+    // Since the cache is *only* for speed, if it actually makes things slower
+    // because of WeakMap collection memory pressure, get rid of the cache.
+    const lastStoreCache = new WeakMap();
+    function finalStore(key) {
+      if (lastStoreCache.has(key) === key) {
+        return lastStoreCache.get(key);
+      }
+      assert(Array.isArray(key), details`trie key must be an array: ${key}`);
+
+      let cursor = root;
+      for (const element of key) {
+        // A "can never happen" that should abort rather than throw
+        assert(element !== Pumpkin, details`internal: Pumpkin leaked!`);
+        if (cursor.has(element)) {
+          // cannot error because of .has test
+          cursor = cursor.get(element);
+        } else {
+          const nextCursor = makeStore(keyName);
+          // cannot error because of .has test
+          cursor.init(element, nextCursor);
+          cursor = nextCursor;
+        }
+      }
+      lastStoreCache.set(key, cursor);
+      return cursor;
+    }
+
+    // /////// Methods /////////
+
+    const has = key => finalStore(key).has(Pumpkin);
+    const init = (key, value) => {
+      const lastStore = finalStore(key);
+      // errors if already there
+      lastStore.init(Pumpkin, value);
+    };
+    const get = key => {
+      const lastStore = finalStore(key);
+      return lastStore.get(Pumpkin);
+    };
```

Comment:
If I'm reading `finalStore` correctly, then `get` and `has` can cause allocation when a key is not found.  Seems to me that probing for things that aren't there oughtn't to be able to make you run out of memory.  I think you need a version of `finalStore` that bails out if it has a miss while walking the key array rather than allocating subtrees to get to the end of the key.

Reply:
Done. But I don't love how I did it --- by adding a flag parameter. If you have a better suggestion, please mention it.

Reply:
I'm not wild about this fix either.  My issue is not the flag but the fact that (a) it does an unnecessary allocation (though it doesn't leak like the prior version did), but more significantly (b) the logic of `has` and `get` is now quite obscure.  It allocates an entire empty map for the sole purpose of it not having a pumpkin entry in it.

It seems to me that `set` and `get`/`has` are fundamentally different operations.  `get` and `has` are about following a chain of maps until you either reach the entry you're looking for or hit a dead end.  `set` is about building a chain of maps until you reach the place where you're going to put something.  I wouldn't use a flag, I'd have two different functions for walking the chain, one that searches and one that builds.

Reply:
It's actually `init` vs `has` vs `get`,`set`,`delete`. Only `init` needs it to not be bound. `has` checks if it is bound. All the others need it to be bound. PTAL (Please Take Another Look)